okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -143,130 +299,339 @@ private static String checkOneDependency(ModuleDescriptor md, InterfaceDescripto
+      return null;
+    }
+    if (checkRequired(modsAvailable, modsEnabled, tml, fix, errors,
+        providedInterfaces, requiredInterfaces, stickyModules)) {
+      return null;
+    }
+    if (checkCompatible(modsAvailable, modsEnabled, tml, fix, errors,
+        providedInterfaces, requiredOptInterfaces, stickyModules)) {
+      return null;
     }
-    return String.join(" ", conflicts);
+    return errors;
   }
 
-  private static TenantModuleDescriptor getNextTM(Map<String, ModuleDescriptor> modsEnabled,
-                                                  List<TenantModuleDescriptor> tml) {
-
-    Iterator<TenantModuleDescriptor> it = tml.iterator();
-    TenantModuleDescriptor tm;
-    while (it.hasNext()) {
-      tm = it.next();
-      TenantModuleDescriptor.Action action = tm.getAction();
-      String id = tm.getId();
-      if (logger.isInfoEnabled()) {
-        logger.info("getNextTM: loop id {} action {}", id, action.name());
+  private static boolean checkMultiple(
+      Map<String, ModuleDescriptor> modsEnabled,
+      List<TenantModuleDescriptor> tml, boolean fix, List<String> errors,
+      Map<String, List<ModuleInterface>> providedInterfaces, Set<String> stickyModules) {
+
+    for (Map.Entry<String, List<ModuleInterface>> entry : providedInterfaces.entrySet()) {
+      if (entry.getValue().size() > 1) {
+        logger.info("Interface {} is defined {} times",
+            entry.getKey(), entry.getValue().size());
+        if (fix) {
+          for (ModuleInterface ent : entry.getValue()) {
+            ModuleDescriptor md = ent.moduleDescriptor;
+            if (!stickyModules.contains(md.getId())) {
+              logger.info("Disable module {}", md.getId());
+              modsEnabled.remove(md.getId());
+              addTenantModule(tml, md.getId(), null, TenantModuleDescriptor.Action.disable);
+              return true;
+            }
+          }
+        }
+        String modules = entry.getValue().stream()
+            .map(x -> x.moduleDescriptor.getId())
+            .collect(Collectors.joining(", "));
+        errors.add(messages.getMessage("10213", modules, entry.getKey()));
       }
-      if (action == TenantModuleDescriptor.Action.enable && !modsEnabled.containsKey(id)) {
-        logger.info("getNextMT: return tm for action enable");
-        return tm;
+    }
+    return false;
+  }
+
+  private static boolean checkRequired(
+      Map<String, ModuleDescriptor> modsAvailable,
+      Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml,
+      boolean fix, List<String> errors, Map<String, List<ModuleInterface>> providedInterfaces,
+      Map<String, List<ModuleInterface>> requiredInterfaces, Set<String> stickyModules) {
+
+    for (Map.Entry<String, List<ModuleInterface>> entry : requiredInterfaces.entrySet()) {
+      List<ModuleInterface> providedModuleInterfaces = providedInterfaces.get(entry.getKey());
+      if (providedModuleInterfaces == null) {
```

Comment:
```
      if (providedInterfaces.get(entry.getKey() != null) {
        continue;  // required interfaces already provided, nothing to do
      }
      logger.info("Interface {} undefined and required", entry.getKey());
```
To reduce indentation and for better readability.

Reply:
done